File: model-optimizer/mo/front/tf/extractors/strided_slice.py
"""
 Copyright (C) 2018-2020 Intel Corporation

 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
"""
import numpy as np

from mo.front.extractor import FrontExtractorOp
from mo.ops.strided_slice import StridedSlice


def int_to_array_bit_mask(im):
    list_repr = list(np.binary_repr(im))
    list_repr.reverse()
    list_repr = [int(li) for li in list_repr]
    return np.array(list_repr, dtype=np.int32)


class StridedSliceFrontExtractor(FrontExtractorOp):
    op = 'StridedSlice'
    enabled = True

    @classmethod
    def extract(cls, node):
        pb = node.pb

        attrs = {
            'begin_mask': int_to_array_bit_mask(pb.attr["begin_mask"].i),
            'end_mask': int_to_array_bit_mask(pb.attr["end_mask"].i),
            'ellipsis_mask': int_to_array_bit_mask(pb.attr["ellipsis_mask"].i),
            'new_axis_mask': int_to_array_bit_mask(pb.attr["new_axis_mask"].i),
            'shrink_axis_mask': int_to_array_bit_mask(pb.attr["shrink_axis_mask"].i),
        }
        attrs['begin_mask'] = np.array([1 - b for b in attrs['begin_mask']], dtype=np.int32)
        attrs['end_mask'] = np.array([1 - b for b in attrs['end_mask']], dtype=np.int32)

        dims = max(map(lambda x: len(x[1]), attrs.items()))

        def extend_mask(in_mask, dims=dims, zeros=True):
            mask = list(in_mask)
            if len(mask) < dims:
                if zeros:
                    mask.extend(np.zeros(dims - len(mask), dtype=np.int32))
                else:
                    mask.extend(np.ones(dims - len(mask), dtype=np.int32))
            return np.array(mask, dtype=np.int32)

        attrs['begin_mask'] = extend_mask(attrs['begin_mask'], zeros=False)
        attrs['end_mask'] = extend_mask(attrs['end_mask'], zeros=False)
        attrs['new_axis_mask'] = extend_mask(attrs['new_axis_mask'])
        attrs['shrink_axis_mask'] = extend_mask(attrs['shrink_axis_mask'])
        attrs['ellipsis_mask'] = extend_mask(attrs['ellipsis_mask'])

        StridedSlice.update_node_stat(node, attrs)
        return cls.enabled

File: ngraph/core/reference/src/runtime/reference/strided_slice.cpp
//*****************************************************************************
// Copyright 2017-2020 Intel Corporation
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//*****************************************************************************

#include <cmath>
#include <stdio.h>

#include "ngraph/check.hpp"
#include "ngraph/runtime/aligned_buffer.hpp"
#include "ngraph/runtime/reference/strided_slice.hpp"

using namespace ngraph;

void runtime::reference::strided_slice(
    const char* arg, char* out, const Shape& arg_shape, const SlicePlan& sp, size_t elem_type)
{
    runtime::AlignedBuffer slice_out_buffer(shape_size(sp.reshape_in_shape) * elem_type);
    slice(reinterpret_cast<const char*>(arg),
          slice_out_buffer.get_ptr<char>(),
          arg_shape,
          Coordinate(sp.begins.begin(), sp.begins.end()),
          Coordinate(sp.ends.begin(), sp.ends.end()),
          Strides(sp.strides.begin(), sp.strides.end()),
          sp.reshape_in_shape,
          elem_type);

    runtime::AlignedBuffer reshape_out_buffer(shape_size(sp.reshape_out_shape) * elem_type);
    opt_kernel::reshape(slice_out_buffer.get_ptr<char>(),
                        reshape_out_buffer.get_ptr<char>(),
                        sp.reshape_in_shape,
                        get_default_order(sp.reshape_in_shape.size()),
                        sp.reshape_out_shape,
                        elem_type);

    reverse(reshape_out_buffer.get_ptr<char>(),
            out,
            sp.reshape_out_shape,
            sp.reshape_out_shape,
            sp.reverse_axes,
            elem_type);
}
